File: test/frame-selection.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { FrameSelection } from "../src/frame-selection.js";
import { ViewSession } from "../src/session.js";
import { layeredCalendarFramesForTest } from "../src/projections.js";

// Direct unit coverage of the class session.js now delegates to. Every
// frame-selection surface (the settings-bar Frame dropdown, the Frames
// panel/dock's leading select and companion checkboxes) reads and writes
// one of these instead of parallel state.

test("a fresh selection defaults its primary to the first id given", () => {
  const selection = new FrameSelection(["calendar:personal", "calendar:work"]);
  assert.equal(selection.primary(), "calendar:personal");
  assert.deepEqual(selection.selected(), ["calendar:personal", "calendar:work"]);
  assert.ok(selection.isPrimary("calendar:personal"));
  assert.ok(!selection.isPrimary("calendar:work"));
});

test("an explicit primary argument wins over the first id, when it is a member", () => {
  const selection = new FrameSelection(["calendar:personal", "calendar:work"], "calendar:work");
  assert.equal(selection.primary(), "calendar:work");
  assert.deepEqual(selection.selected(), ["calendar:work", "calendar:personal"], "primary sorts first");
});

test("a primary argument that is not a member of the ids is ignored", () => {
  const selection = new FrameSelection(["calendar:personal"], "calendar:nowhere");
  assert.equal(selection.primary(), "calendar:personal");
});

test("toggle adds an unselected id and removes a selected one", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.toggle("calendar:work");
  assert.ok(selection.isSelected("calendar:work"));
  assert.equal(selection.primary(), "calendar:personal", "adding a companion never moves the primary");
  selection.toggle("calendar:work");
  assert.ok(!selection.isSelected("calendar:work"));
});

test("toggle refuses to remove the last remaining id", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.toggle("calendar:personal");
  assert.deepEqual(selection.selected(), ["calendar:personal"]);
});

test("toggling the primary out promotes the next selected id, never leaving the selection empty", () => {
  const selection = new FrameSelection(["calendar:a", "calendar:b", "calendar:c"]);
  selection.toggle("calendar:a");
  assert.equal(selection.primary(), "calendar:b", "the id that followed the removed primary is promoted");
  assert.deepEqual(new Set(selection.selected()), new Set(["calendar:b", "calendar:c"]));
});

test("toggling out the last-positioned primary wraps to the new first id", () => {
  const selection = new FrameSelection(["calendar:a", "calendar:b"], "calendar:b");
  selection.toggle("calendar:b");
  assert.equal(selection.primary(), "calendar:a");
});

test("setPrimary reassigns the marker without changing membership — the 8.19 regression, at the class level", () => {
  const selection = new FrameSelection(["calendar:personal", "calendar:work", "calendar:family"]);
  selection.setPrimary("calendar:family");
  assert.equal(selection.primary(), "calendar:family");
  assert.deepEqual(new Set(selection.selected()), new Set(["calendar:personal", "calendar:work", "calendar:family"]),
    "reassigning the primary must not drop any previously selected frame");
});

test("setPrimary on an id outside the selection adds it rather than replacing the selection", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.setPrimary("calendar:work");
  assert.equal(selection.primary(), "calendar:work");
  assert.ok(selection.isSelected("calendar:personal"), "the previous primary is demoted, not dropped");
});

test("setSelection bulk-replaces, keeping the previous primary when it survives", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.setSelection(["calendar:work", "calendar:personal", "calendar:family"]);
  assert.equal(selection.primary(), "calendar:personal");
  selection.setSelection(["calendar:work", "calendar:family"]);
  assert.equal(selection.primary(), "calendar:work", "the previous primary dropped out, so the first offered id leads");
});

test("setSelection refuses an empty list", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.setSelection([]);
  assert.deepEqual(selection.selected(), ["calendar:personal"]);
});

test("prune drops ids that no longer exist and reassigns a stale primary", () => {
  const selection = new FrameSelection(["calendar:personal", "calendar:work", "calendar:family"]);
  selection.prune(["calendar:work", "calendar:family"]);
  assert.equal(selection.primary(), "calendar:work");
  assert.deepEqual(new Set(selection.selected()), new Set(["calendar:work", "calendar:family"]));
});

test("prune falls back to an arbitrary valid id when nothing of the old selection survives", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.prune(["calendar:work", "calendar:family"]);
  assert.equal(selection.primary(), "calendar:work");
});

test("prune against an empty valid set leaves the selection untouched", () => {
  const selection = new FrameSelection(["calendar:personal"]);
  selection.prune([]);
  assert.deepEqual(selection.selected(), ["calendar:personal"], "no frames loaded yet is not proof the selection is wrong");
});

// The actual regression, at the render boundary: layeredCalendarFrames used
// to read only frames[leading].display.overlays, so checking a companion in
// the toolbar's Frame dropdown or the Frames panel wrote session state the
// renderer never consumed — "the highest frame selected acts as the view
// frame, lower selected frames do not overlay." This test fails against the
// pre-fix implementation and passes now that the projection reads the one
// selection.
test("every selected frame overlays in the calendar projection, primary first", () => {
  const document = {
    frames: {
      "calendar:personal": { id: "calendar:personal", title: "Personal", traits: ["calendar"] },
      "calendar:work": { id: "calendar:work", title: "Work", traits: ["calendar"] },
      "calendar:family": { id: "calendar:family", title: "Family", traits: ["calendar"] }
    }
  };
  const session = new ViewSession({ activeFrame: "calendar:personal" });
  session.setFrameSelection(["calendar:personal", "calendar:work", "calendar:family"]);
  const layered = layeredCalendarFramesForTest({ document, session }, session.activeFrame);
  assert.deepEqual(layered.map((frame) => frame.id), ["calendar:personal", "calendar:work", "calendar:family"]);
});

test("display.overlays plays no role in the projection any more", () => {
  const document = {
    frames: {
      "calendar:personal": { id: "calendar:personal", title: "Personal", traits: ["calendar"], display: { overlays: ["calendar:family"] } },
      "calendar:family": { id: "calendar:family", title: "Family", traits: ["calendar"] }
    }
  };
  const session = new ViewSession({ activeFrame: "calendar:personal" });
  const layered = layeredCalendarFramesForTest({ document, session }, session.activeFrame);
  assert.deepEqual(layered.map((frame) => frame.id), ["calendar:personal"],
    "a stale display.overlays entry that was never selected must not appear");
});

test("requesting a non-primary frame explicitly renders it alone", () => {
  const document = {
    frames: {
      "calendar:personal": { id: "calendar:personal", title: "Personal", traits: ["calendar"] },
      "calendar:work": { id: "calendar:work", title: "Work", traits: ["calendar"] }
    }
  };
  const session = new ViewSession({ activeFrame: "calendar:personal" });
  session.setFrameSelection(["calendar:personal", "calendar:work"]);
  const layered = layeredCalendarFramesForTest({ document, session }, "calendar:work");
  assert.deepEqual(layered.map((frame) => frame.id), ["calendar:work"]);
});

File: app/test/core/frame_selection_test.dart
// Frame selection: ordered ids plus an explicit primary.
//
// The 8.19 field report was that reassigning the leading frame silently dropped
// every companion. So the spec is not a list of scenarios -- it is the four
// invariants, checked after every operation of a RANDOM operation sequence, which
// is the only way to say "no sequence of clicks can break this".

import 'dart:math';

import 'package:chronolog/core/frame_selection.dart';
import 'package:test/test.dart';

import 'corpus.dart';

List<String> _pool(int size) => [for (var i = 0; i < size; i++) 'calendar:$i'];

/// The four invariants, together, after every single operation.
void _holds(FrameSelection selection, String after) {
  final selected = selection.selected();
  expect(selected, isNotEmpty, reason: 'never empty: $after');
  expect(
    selection.primary(),
    isNotNull,
    reason: 'a non-empty selection always leads with something: $after',
  );
  expect(selected, contains(selection.primary()), reason: 'the primary is a member: $after');
  expect(selected.first, selection.primary(), reason: 'primary first: $after');
  expect(selected.toSet(), hasLength(selected.length), reason: 'no duplicates: $after');
  for (final id in selected) {
    expect(selection.isSelected(id), isTrue, reason: after);
  }
  expect(selection.isPrimary(selection.primary()!), isTrue, reason: after);
}

void main() {
  test('random operation sequences: every invariant holds throughout', () {
    for (final seed in seeds(60)) {
      final random = Random(seed);
      final pool = _pool(2 + random.nextInt(6));
      final selection = FrameSelection([pool.first]);
      _holds(selection, 'construction');
      for (var step = 0; step < 30; step++) {
        final id = pool[random.nextInt(pool.length)];
        final before = selection.selected().toSet();
        switch (random.nextInt(4)) {
          case 0:
            selection.toggle(id);
            _holds(selection, 'seed $seed step $step toggle $id');
          case 1:
            selection.setPrimary(id);
            _holds(selection, 'seed $seed step $step setPrimary $id');
            expect(selection.primary(), id, reason: 'setPrimary always lands the marker');
            // MEMBERSHIP IS NEVER REDUCED by moving the marker. This is the
            // 8.19 regression, stated as an invariant.
            expect(
              selection.selected().toSet(),
              containsAll(before),
              reason: 'seed $seed step $step: setPrimary dropped a frame',
            );
          case 2:
            final wanted = [
              for (final candidate in pool)
                if (random.nextBool()) candidate,
            ];
            selection.setSelection(wanted);
            _holds(selection, 'seed $seed step $step setSelection $wanted');
            if (wanted.isEmpty) {
              expect(
                selection.selected().toSet(),
                before,
                reason: 'an empty list is refused outright',
              );
            } else {
              expect(selection.selected().toSet(), wanted.toSet());
            }
          default:
            final valid = [
              for (final candidate in pool)
                if (random.nextBool()) candidate,
            ];
            selection.prune(valid);
            _holds(selection, 'seed $seed step $step prune $valid');
            if (valid.isEmpty) {
              expect(
                selection.selected().toSet(),
                before,
                reason: 'no frames loaded yet is not proof the selection is wrong',
              );
            } else {
              expect(
                selection.selected().every(valid.contains),
                isTrue,
                reason: 'seed $seed step $step',
              );
            }
        }
      }
    }
  });

  test('the last remaining id can never be toggled away', () {
    for (final seed in seeds(20)) {
      final random = Random(seed);
      final pool = _pool(1 + random.nextInt(5));
      final selection = FrameSelection(pool);
      // Toggle everything off, in a random order, as many times as it takes.
      for (var step = 0; step < pool.length * 3; step++) {
        selection.toggle(pool[random.nextInt(pool.length)]);
        _holds(selection, 'seed $seed');
      }
      expect(selection.selected(), isNotEmpty);
    }
  });

  group('ruled anchors', () {
    test('a fresh selection leads with the first id given', () {
      final selection = FrameSelection(['calendar:a', 'calendar:b']);
      expect(selection.primary(), 'calendar:a');
      expect(selection.selected(), ['calendar:a', 'calendar:b']);
    });

    test('an explicit primary wins, but only when it is a member', () {
      expect(FrameSelection(['calendar:a', 'calendar:b'], 'calendar:b').selected(), [
        'calendar:b',
        'calendar:a',
      ]);
      expect(FrameSelection(['calendar:a'], 'calendar:nowhere').primary(), 'calendar:a');
    });

    test('adding a companion never moves the marker', () {
      final selection = FrameSelection(['calendar:a']);
      selection.toggle('calendar:b');
      expect(selection.isSelected('calendar:b'), isTrue);
      expect(selection.primary(), 'calendar:a');
    });

    test('toggling the primary out promotes the id that followed it', () {
      final selection = FrameSelection(['calendar:a', 'calendar:b', 'calendar:c']);
      selection.toggle('calendar:a');
      expect(selection.primary(), 'calendar:b');
      expect(selection.selected().toSet(), {'calendar:b', 'calendar:c'});
    });

    test('toggling out the last-positioned primary wraps to the new first', () {
      final selection = FrameSelection(['calendar:a', 'calendar:b'], 'calendar:b');
      selection.toggle('calendar:b');
      expect(selection.primary(), 'calendar:a');
    });

    test('RULED ANCHOR: setPrimary reassigns the marker, and nothing else', () {
      final selection = FrameSelection(['calendar:a', 'calendar:b', 'calendar:c']);
      selection.setPrimary('calendar:c');
      expect(selection.primary(), 'calendar:c');
      expect(selection.selected().toSet(), {'calendar:a', 'calendar:b', 'calendar:c'});
    });

    test('setPrimary on an outsider adds it rather than replacing', () {
      final selection = FrameSelection(['calendar:a']);
      selection.setPrimary('calendar:b');
      expect(selection.primary(), 'calendar:b');
      expect(selection.isSelected('calendar:a'), isTrue);
    });

    test('setSelection keeps a surviving primary, otherwise the first offered', () {
      final selection = FrameSelection(['calendar:a']);
      selection.setSelection(['calendar:b', 'calendar:a', 'calendar:c']);
      expect(selection.primary(), 'calendar:a');
      selection.setSelection(['calendar:b', 'calendar:c']);
      expect(selection.primary(), 'calendar:b');
    });

    test('prune reassigns a stale primary, and falls back only if it must', () {
      final wide = FrameSelection(['calendar:a', 'calendar:b', 'calendar:c']);
      wide.prune(['calendar:b', 'calendar:c']);
      expect(wide.primary(), 'calendar:b');
      final orphan = FrameSelection(['calendar:a']);
      orphan.prune(['calendar:b', 'calendar:c']);
      expect(orphan.primary(), 'calendar:b');
    });

    test('an empty or blank id is not a selection', () {
      final selection = FrameSelection(['', 'calendar:a', '']);
      expect(selection.selected(), ['calendar:a']);
      selection.toggle('');
      selection.setPrimary('');
      expect(selection.selected(), ['calendar:a']);
    });
  });
}
